File: Java/CesiumLanguageWriterTests/translatedSrc/cesiumlanguagewritertests/TestLabelCesiumWriter.java
package cesiumlanguagewritertests;


import agi.foundation.compatibility.*;
import agi.foundation.compatibility.AssertHelper;
import agi.foundation.compatibility.ColorHelper;
import agi.foundation.compatibility.MapHelper;
import agi.foundation.compatibility.TestContextRule;
import agi.foundation.compatibility.Using;
import cesiumlanguagewriter.*;
import cesiumlanguagewriter.advanced.*;
import java.awt.Color;
import java.io.StringWriter;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Map;
import javax.annotation.Nonnull;
import org.junit.ClassRule;
import org.junit.FixMethodOrder;
import org.junit.Rule;
import org.junit.runners.MethodSorters;
import org.junit.Test;

@SuppressWarnings({
    "unused",
    "deprecation",
    "serial"
})
@FixMethodOrder(MethodSorters.NAME_ASCENDING)
public class TestLabelCesiumWriter extends TestCesiumPropertyWriter<LabelCesiumWriter> {
    @Test
    public final void testShow() {
        final boolean expected = true;
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<LabelCesiumWriter> using$1 = new Using<LabelCesiumWriter>(packet.openLabelProperty())) {
                final LabelCesiumWriter label = using$1.resource;
                try (Using<LabelCesiumWriter> using$2 = new Using<LabelCesiumWriter>(label.openInterval())) {
                    final LabelCesiumWriter interval = using$2.resource;
                    interval.writeShowProperty(expected);
                }
            }
        }
        assertPropertyJson(LabelCesiumWriter.ShowPropertyName, expected);
    }

    @Test
    public final void testFont() {
        final String expected = "30px Helvetica";
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<LabelCesiumWriter> using$1 = new Using<LabelCesiumWriter>(packet.openLabelProperty())) {
                final LabelCesiumWriter label = using$1.resource;
                try (Using<LabelCesiumWriter> using$2 = new Using<LabelCesiumWriter>(label.openInterval())) {
                    final LabelCesiumWriter interval = using$2.resource;
                    interval.writeFontProperty(expected);
                }
            }
        }
        assertPropertyJson(LabelCesiumWriter.FontPropertyName, expected);
    }

    @Test
    public final void testStyle() {
        final CesiumLabelStyle expected = CesiumLabelStyle.FILL_AND_OUTLINE;
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<LabelCesiumWriter> using$1 = new Using<LabelCesiumWriter>(packet.openLabelProperty())) {
                final LabelCesiumWriter label = using$1.resource;
                try (Using<LabelCesiumWriter> using$2 = new Using<LabelCesiumWriter>(label.openInterval())) {
                    final LabelCesiumWriter interval = using$2.resource;
                    interval.writeStyleProperty(expected);
                }
            }
        }
        assertPropertyJson(LabelCesiumWriter.StylePropertyName, expected);
    }

    @Test
    public final void testScale() {
        final double expected = 2.0;
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<LabelCesiumWriter> using$1 = new Using<LabelCesiumWriter>(packet.openLabelProperty())) {
                final LabelCesiumWriter label = using$1.resource;
                try (Using<LabelCesiumWriter> using$2 = new Using<LabelCesiumWriter>(label.openInterval())) {
                    final LabelCesiumWriter interval = using$2.resource;
                    interval.writeScaleProperty(expected);
                }
            }
        }
        assertPropertyJson(LabelCesiumWriter.ScalePropertyName, expected);
    }

    @Test
    public final void testPixelOffset() {
        Rectangular expected = new Rectangular(12.5, 3.5);
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<LabelCesiumWriter> using$1 = new Using<LabelCesiumWriter>(packet.openLabelProperty())) {
                final LabelCesiumWriter label = using$1.resource;
                try (Using<LabelCesiumWriter> using$2 = new Using<LabelCesiumWriter>(label.openInterval())) {
                    final LabelCesiumWriter interval = using$2.resource;
                    interval.writePixelOffsetProperty(expected);
                }
            }
        }
        assertPropertyJson(LabelCesiumWriter.PixelOffsetPropertyName, expected);
    }

    @Test
    public final void testEyeOffset() {
        Cartesian expected = new Cartesian(1.0, 2.0, 3.0);
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<LabelCesiumWriter> using$1 = new Using<LabelCesiumWriter>(packet.openLabelProperty())) {
                final LabelCesiumWriter label = using$1.resource;
                try (Using<LabelCesiumWriter> using$2 = new Using<LabelCesiumWriter>(label.openInterval())) {
                    final LabelCesiumWriter interval = using$2.resource;
                    interval.writeEyeOffsetProperty(expected);
                }
            }
        }
        assertPropertyJson(LabelCesiumWriter.EyeOffsetPropertyName, expected);
    }

    @Test
    public final void testHorizontalOrigin() {
        final CesiumHorizontalOrigin expected = CesiumHorizontalOrigin.CENTER;
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<LabelCesiumWriter> using$1 = new Using<LabelCesiumWriter>(packet.openLabelProperty())) {
                final LabelCesiumWriter label = using$1.resource;
                try (Using<LabelCesiumWriter> using$2 = new Using<LabelCesiumWriter>(label.openInterval())) {
                    final LabelCesiumWriter interval = using$2.resource;
                    interval.writeHorizontalOriginProperty(expected);
                }
            }
        }
        assertPropertyJson(LabelCesiumWriter.HorizontalOriginPropertyName, expected);
    }

    @Test
    public final void testVerticalOrigin() {
        final CesiumVerticalOrigin expected = CesiumVerticalOrigin.CENTER;
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<LabelCesiumWriter> using$1 = new Using<LabelCesiumWriter>(packet.openLabelProperty())) {
                final LabelCesiumWriter label = using$1.resource;
                try (Using<LabelCesiumWriter> using$2 = new Using<LabelCesiumWriter>(label.openInterval())) {
                    final LabelCesiumWriter interval = using$2.resource;
                    interval.writeVerticalOriginProperty(expected);
                }
            }
        }
        assertPropertyJson(LabelCesiumWriter.VerticalOriginPropertyName, expected);
    }

    @Test
    public final void testHeightReference() {
        final CesiumHeightReference expected = CesiumHeightReference.CLAMP_TO_GROUND;
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<LabelCesiumWriter> using$1 = new Using<LabelCesiumWriter>(packet.openLabelProperty())) {
                final LabelCesiumWriter label = using$1.resource;
                try (Using<LabelCesiumWriter> using$2 = new Using<LabelCesiumWriter>(label.openInterval())) {
                    final LabelCesiumWriter interval = using$2.resource;
                    interval.writeHeightReferenceProperty(expected);
                }
            }
        }
        assertPropertyJson(LabelCesiumWriter.HeightReferencePropertyName, expected);
    }

    @Test
    public final void testFillColor() {
        Color expected = Color.BLUE;
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<LabelCesiumWriter> using$1 = new Using<LabelCesiumWriter>(packet.openLabelProperty())) {
                final LabelCesiumWriter label = using$1.resource;
                try (Using<LabelCesiumWriter> using$2 = new Using<LabelCesiumWriter>(label.openInterval())) {
                    final LabelCesiumWriter interval = using$2.resource;
                    interval.writeFillColorProperty(expected);
                }
            }
        }
        assertPropertyJson(LabelCesiumWriter.FillColorPropertyName, expected);
    }

    @Test
    public final void testOutlineColor() {
        Color expected = ColorHelper.GREEN;
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<LabelCesiumWriter> using$1 = new Using<LabelCesiumWriter>(packet.openLabelProperty())) {
                final LabelCesiumWriter label = using$1.resource;
                try (Using<LabelCesiumWriter> using$2 = new Using<LabelCesiumWriter>(label.openInterval())) {
                    final LabelCesiumWriter interval = using$2.resource;
                    interval.writeOutlineColorProperty(expected);
                }
            }
        }
        assertPropertyJson(LabelCesiumWriter.OutlineColorPropertyName, expected);
    }

    @Test
    public final void testOutlineWidth() {
        final double expected = 2.0;
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<LabelCesiumWriter> using$1 = new Using<LabelCesiumWriter>(packet.openLabelProperty())) {
                final LabelCesiumWriter label = using$1.resource;
                try (Using<LabelCesiumWriter> using$2 = new Using<LabelCesiumWriter>(label.openInterval())) {
                    final LabelCesiumWriter interval = using$2.resource;
                    interval.writeOutlineWidthProperty(expected);
                }
            }
        }
        assertPropertyJson(LabelCesiumWriter.OutlineWidthPropertyName, expected);
    }

    @Test
    public final void testPixelOffsetScaleByDistance() {
        NearFarScalar expected = new NearFarScalar(10.0, 1.0, 20.0, 2.0);
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<LabelCesiumWriter> using$1 = new Using<LabelCesiumWriter>(packet.openLabelProperty())) {
                final LabelCesiumWriter label = using$1.resource;
                try (Using<LabelCesiumWriter> using$2 = new Using<LabelCesiumWriter>(label.openInterval())) {
                    final LabelCesiumWriter interval = using$2.resource;
                    interval.writePixelOffsetScaleByDistanceProperty(expected);
                }
            }
        }
        assertPropertyJson(LabelCesiumWriter.PixelOffsetScaleByDistancePropertyName, expected);
    }

    @Test
    public final void testDistanceDisplayCondition() {
        Bounds expected = new Bounds(10.0, 20.0);
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<LabelCesiumWriter> using$1 = new Using<LabelCesiumWriter>(packet.openLabelProperty())) {
                final LabelCesiumWriter label = using$1.resource;
                try (Using<LabelCesiumWriter> using$2 = new Using<LabelCesiumWriter>(label.openInterval())) {
                    final LabelCesiumWriter interval = using$2.resource;
                    interval.writeDistanceDisplayConditionProperty(expected);
                }
            }
        }
        assertPropertyJson(LabelCesiumWriter.DistanceDisplayConditionPropertyName, expected);
    }

    @Test
    public final void testDisableDepthTestDistance() {
        final double expected = 1234.0;
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<LabelCesiumWriter> using$1 = new Using<LabelCesiumWriter>(packet.openLabelProperty())) {
                final LabelCesiumWriter label = using$1.resource;
                try (Using<LabelCesiumWriter> using$2 = new Using<LabelCesiumWriter>(label.openInterval())) {
                    final LabelCesiumWriter interval = using$2.resource;
                    interval.writeDisableDepthTestDistanceProperty(expected);
                }
            }
        }
        assertPropertyJson(LabelCesiumWriter.DisableDepthTestDistancePropertyName, expected);
    }

    @Test
    public final void testScaleByDistance() {
        NearFarScalar expected = new NearFarScalar(100.5, 1.5, 200.5, 2.5);
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<LabelCesiumWriter> using$1 = new Using<LabelCesiumWriter>(packet.openLabelProperty())) {
                final LabelCesiumWriter label = using$1.resource;
                try (Using<LabelCesiumWriter> using$2 = new Using<LabelCesiumWriter>(label.openInterval())) {
                    final LabelCesiumWriter interval = using$2.resource;
                    interval.writeScaleByDistanceProperty(expected);
                }
            }
        }
        assertPropertyJson(LabelCesiumWriter.ScaleByDistancePropertyName, expected);
    }

    @Test
    public final void testText() {
        final String expected = "asdf";
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<LabelCesiumWriter> using$1 = new Using<LabelCesiumWriter>(packet.openLabelProperty())) {
                final LabelCesiumWriter label = using$1.resource;
                try (Using<LabelCesiumWriter> using$2 = new Using<LabelCesiumWriter>(label.openInterval())) {
                    final LabelCesiumWriter interval = using$2.resource;
                    interval.writeTextProperty(expected);
                }
            }
        }
        assertPropertyJson(LabelCesiumWriter.TextPropertyName, expected);
    }

    @Test
    public final void testTextInterval() {
        JulianDate startDate = new GregorianDate(2012, 4, 2, 12, 0, 0.0).toJulianDate();
        JulianDate stopDate = new GregorianDate(2012, 4, 2, 12, 1, 0.0).toJulianDate();
        JulianDate interval1Start = startDate;
        JulianDate interval1Stop = startDate.addSeconds(1.0);
        JulianDate interval2Start = interval1Stop;
        JulianDate interval2Stop = startDate.addSeconds(2.0);
        JulianDate interval3Start = interval2Stop;
        JulianDate interval3Stop = stopDate;
        final String interval1Value = "A";
        final String interval2Value = "B";
        final String interval3Value = "C";
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<LabelCesiumWriter> using$1 = new Using<LabelCesiumWriter>(packet.openLabelProperty())) {
                final LabelCesiumWriter label = using$1.resource;
                try (Using<StringCesiumWriter> using$2 = new Using<StringCesiumWriter>(label.openTextProperty())) {
                    final StringCesiumWriter text = using$2.resource;
                    try (Using<CesiumIntervalListWriter<StringCesiumWriter>> using$3 = new Using<CesiumIntervalListWriter<StringCesiumWriter>>(text.openMultipleIntervals())) {
                        final CesiumIntervalListWriter<StringCesiumWriter> textIntervals = using$3.resource;
                        try (Using<StringCesiumWriter> using$4 = new Using<StringCesiumWriter>(textIntervals.openInterval(interval1Start, interval1Stop))) {
                            final StringCesiumWriter interval = using$4.resource;
                            interval.writeString(interval1Value);
                        }
                        try (Using<StringCesiumWriter> using$5 = new Using<StringCesiumWriter>(textIntervals.openInterval(interval2Start, interval2Stop))) {
                            final StringCesiumWriter interval = using$5.resource;
                            interval.writeString(interval2Value);
                        }
                        try (Using<StringCesiumWriter> using$6 = new Using<StringCesiumWriter>(textIntervals.openInterval(interval3Start, interval3Stop))) {
                            final StringCesiumWriter interval = using$6.resource;
                            interval.writeString(interval3Value);
                        }
                    }
                }
            }
        }
        final Map<String, Object> tempCollection$1 = MapHelper.create();
        MapHelper.add(tempCollection$1, "interval", CesiumFormattingHelper.toIso8601Interval(interval1Start, interval1Stop, Iso8601Format.COMPACT));
        MapHelper.add(tempCollection$1, StringCesiumWriter.StringPropertyName, interval1Value);
        final Map<String, Object> tempCollection$2 = MapHelper.create();
        MapHelper.add(tempCollection$2, "interval", CesiumFormattingHelper.toIso8601Interval(interval2Start, interval2Stop, Iso8601Format.COMPACT));
        MapHelper.add(tempCollection$2, StringCesiumWriter.StringPropertyName, interval2Value);
        final Map<String, Object> tempCollection$3 = MapHelper.create();
        MapHelper.add(tempCollection$3, "interval", CesiumFormattingHelper.toIso8601Interval(interval3Start, interval3Stop, Iso8601Format.COMPACT));
        MapHelper.add(tempCollection$3, StringCesiumWriter.StringPropertyName, interval3Value);
        final ArrayList<Map<String, Object>> tempCollection$0 = new ArrayList<Map<String, Object>>();
        tempCollection$0.add(tempCollection$1);
        tempCollection$0.add(tempCollection$2);
        tempCollection$0.add(tempCollection$3);
        assertPropertyJson(LabelCesiumWriter.TextPropertyName, tempCollection$0);
    }

    @Test
    public final void testShowBackground() {
        final boolean expected = true;
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<LabelCesiumWriter> using$1 = new Using<LabelCesiumWriter>(packet.openLabelProperty())) {
                final LabelCesiumWriter label = using$1.resource;
                try (Using<LabelCesiumWriter> using$2 = new Using<LabelCesiumWriter>(label.openInterval())) {
                    final LabelCesiumWriter interval = using$2.resource;
                    interval.writeShowBackgroundProperty(expected);
                }
            }
        }
        assertPropertyJson(LabelCesiumWriter.ShowBackgroundPropertyName, expected);
    }

    @Test
    public final void testBackgroundColor() {
        Color expected = Color.RED;
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<LabelCesiumWriter> using$1 = new Using<LabelCesiumWriter>(packet.openLabelProperty())) {
                final LabelCesiumWriter label = using$1.resource;
                try (Using<LabelCesiumWriter> using$2 = new Using<LabelCesiumWriter>(label.openInterval())) {
                    final LabelCesiumWriter interval = using$2.resource;
                    interval.writeBackgroundColorProperty(expected);
                }
            }
        }
        assertPropertyJson(LabelCesiumWriter.BackgroundColorPropertyName, expected);
    }

    @Test
    public final void testBackgroundPadding() {
        Rectangular expected = new Rectangular(3.0, 4.0);
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<LabelCesiumWriter> using$1 = new Using<LabelCesiumWriter>(packet.openLabelProperty())) {
                final LabelCesiumWriter label = using$1.resource;
                try (Using<LabelCesiumWriter> using$2 = new Using<LabelCesiumWriter>(label.openInterval())) {
                    final LabelCesiumWriter interval = using$2.resource;
                    interval.writeBackgroundPaddingProperty(expected);
                }
            }
        }
        assertPropertyJson(LabelCesiumWriter.BackgroundPaddingPropertyName, expected);
    }

    @Test
    public final void testTranslucencyByDistance() {
        NearFarScalar expected = new NearFarScalar(100.0, 1.0, 200.0, 2.0);
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<LabelCesiumWriter> using$1 = new Using<LabelCesiumWriter>(packet.openLabelProperty())) {
                final LabelCesiumWriter label = using$1.resource;
                try (Using<LabelCesiumWriter> using$2 = new Using<LabelCesiumWriter>(label.openInterval())) {
                    final LabelCesiumWriter interval = using$2.resource;
                    interval.writeTranslucencyByDistanceProperty(expected);
                }
            }
        }
        assertPropertyJson(LabelCesiumWriter.TranslucencyByDistancePropertyName, expected);
    }

    @Test
    public final void testTranslucencyByDistanceSamples() {
        JulianDate epoch = new GregorianDate(2012, 4, 2, 12, 0, 0.0).toJulianDate();
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<LabelCesiumWriter> using$1 = new Using<LabelCesiumWriter>(packet.openLabelProperty())) {
                final LabelCesiumWriter label = using$1.resource;
                try (Using<LabelCesiumWriter> using$2 = new Using<LabelCesiumWriter>(label.openInterval())) {
                    final LabelCesiumWriter interval = using$2.resource;
                    try (Using<NearFarScalarCesiumWriter> using$3 = new Using<NearFarScalarCesiumWriter>(interval.openTranslucencyByDistanceProperty())) {
                        final NearFarScalarCesiumWriter scaleByDistance = using$3.resource;
                        ArrayList<JulianDate> dates = new ArrayList<JulianDate>();
                        ArrayList<NearFarScalar> values = new ArrayList<NearFarScalar>();
                        dates.add(epoch);
                        values.add(new NearFarScalar(100.0, 1.0, 200.0, 2.0));
                        dates.add(epoch.addSeconds(60.0));
                        values.add(new NearFarScalar(200.0, 1.0, 300.0, 2.0));
                        scaleByDistance.writeNearFarScalar(dates, values);
                    }
                }
            }
        }
        AssertHelper.assertEquals("{\"label\":{\"translucencyByDistance\":{\"epoch\":\"20120402T12Z\",\"nearFarScalar\":[0,100,1,200,2,60,200,1,300,2]}}}", getStringWriter().toString());
    }

    private final void assertPropertyJson(String propertyName, Object value) {
        final Map<String, Object> tempCollection$0 = MapHelper.create();
        MapHelper.add(tempCollection$0, propertyName, value);
        assertExpectedJson(PacketCesiumWriter.LabelPropertyName, tempCollection$0);
    }

    @Override
    protected CesiumPropertyWriter<LabelCesiumWriter> createPropertyWriter(String propertyName) {
        return new LabelCesiumWriter(propertyName);
    }

    @Nonnull
    private static final TestContextRule rule$testContext = new TestContextRule();

    @Nonnull
    @Rule
    @ClassRule
    public static TestContextRule getRule$testContext() {
        return rule$testContext;
    }
}

File: DotNet/CesiumLanguageWriterTests/TestLabelCesiumWriter.cs
using System.Collections.Generic;
using System.Drawing;
using CesiumLanguageWriter;
using CesiumLanguageWriter.Advanced;
using NUnit.Framework;

namespace CesiumLanguageWriterTests
{
    [TestFixture]
    public class TestLabelCesiumWriter : TestCesiumPropertyWriter<LabelCesiumWriter>
    {
        [Test]
        public void TestShow()
        {
            const bool expected = true;

            using (var packet = OpenPacket())
            using (var label = packet.OpenLabelProperty())
            using (var interval = label.OpenInterval())
            {
                interval.WriteShowProperty(expected);
            }

            AssertPropertyJson(LabelCesiumWriter.ShowPropertyName, expected);
        }

        [Test]
        public void TestFont()
        {
            const string expected = "30px Helvetica";

            using (var packet = OpenPacket())
            using (var label = packet.OpenLabelProperty())
            using (var interval = label.OpenInterval())
            {
                interval.WriteFontProperty(expected);
            }

            AssertPropertyJson(LabelCesiumWriter.FontPropertyName, expected);
        }

        [Test]
        public void TestStyle()
        {
            const CesiumLabelStyle expected = CesiumLabelStyle.FillAndOutline;

            using (var packet = OpenPacket())
            using (var label = packet.OpenLabelProperty())
            using (var interval = label.OpenInterval())
            {
                interval.WriteStyleProperty(expected);
            }

            AssertPropertyJson(LabelCesiumWriter.StylePropertyName, expected);
        }

        [Test]
        public void TestScale()
        {
            const double expected = 2.0;

            using (var packet = OpenPacket())
            using (var label = packet.OpenLabelProperty())
            using (var interval = label.OpenInterval())
            {
                interval.WriteScaleProperty(expected);
            }

            AssertPropertyJson(LabelCesiumWriter.ScalePropertyName, expected);
        }

        [Test]
        public void TestPixelOffset()
        {
            var expected = new Rectangular(12.5, 3.5);

            using (var packet = OpenPacket())
            using (var label = packet.OpenLabelProperty())
            using (var interval = label.OpenInterval())
            {
                interval.WritePixelOffsetProperty(expected);
            }

            AssertPropertyJson(LabelCesiumWriter.PixelOffsetPropertyName, expected);
        }

        [Test]
        public void TestEyeOffset()
        {
            var expected = new Cartesian(1.0, 2.0, 3.0);

            using (var packet = OpenPacket())
            using (var label = packet.OpenLabelProperty())
            using (var interval = label.OpenInterval())
            {
                interval.WriteEyeOffsetProperty(expected);
            }

            AssertPropertyJson(LabelCesiumWriter.EyeOffsetPropertyName, expected);
        }

        [Test]
        public void TestHorizontalOrigin()
        {
            const CesiumHorizontalOrigin expected = CesiumHorizontalOrigin.Center;

            using (var packet = OpenPacket())
            using (var label = packet.OpenLabelProperty())
            using (var interval = label.OpenInterval())
            {
                interval.WriteHorizontalOriginProperty(expected);
            }

            AssertPropertyJson(LabelCesiumWriter.HorizontalOriginPropertyName, expected);
        }

        [Test]
        public void TestVerticalOrigin()
        {
            const CesiumVerticalOrigin expected = CesiumVerticalOrigin.Center;

            using (var packet = OpenPacket())
            using (var label = packet.OpenLabelProperty())
            using (var interval = label.OpenInterval())
            {
                interval.WriteVerticalOriginProperty(expected);
            }

            AssertPropertyJson(LabelCesiumWriter.VerticalOriginPropertyName, expected);
        }

        [Test]
        public void TestHeightReference()
        {
            const CesiumHeightReference expected = CesiumHeightReference.ClampToGround;

            using (var packet = OpenPacket())
            using (var label = packet.OpenLabelProperty())
            using (var interval = label.OpenInterval())
            {
                interval.WriteHeightReferenceProperty(expected);
            }

            AssertPropertyJson(LabelCesiumWriter.HeightReferencePropertyName, expected);
        }

        [Test]
        public void TestFillColor()
        {
            var expected = Color.Blue;

            using (var packet = OpenPacket())
            using (var label = packet.OpenLabelProperty())
            using (var interval = label.OpenInterval())
            {
                interval.WriteFillColorProperty(expected);
            }

            AssertPropertyJson(LabelCesiumWriter.FillColorPropertyName, expected);
        }

        [Test]
        public void TestOutlineColor()
        {
            var expected = Color.Green;

            using (var packet = OpenPacket())
            using (var label = packet.OpenLabelProperty())
            using (var interval = label.OpenInterval())
            {
                interval.WriteOutlineColorProperty(expected);
            }

            AssertPropertyJson(LabelCesiumWriter.OutlineColorPropertyName, expected);
        }

        [Test]
        public void TestOutlineWidth()
        {
            const double expected = 2.0;

            using (var packet = OpenPacket())
            using (var label = packet.OpenLabelProperty())
            using (var interval = label.OpenInterval())
            {
                interval.WriteOutlineWidthProperty(expected);
            }

            AssertPropertyJson(LabelCesiumWriter.OutlineWidthPropertyName, expected);
        }

        [Test]
        public void TestPixelOffsetScaleByDistance()
        {
            var expected = new NearFarScalar(10.0, 1.0, 20.0, 2.0);

            using (var packet = OpenPacket())
            using (var label = packet.OpenLabelProperty())
            using (var interval = label.OpenInterval())
            {
                interval.WritePixelOffsetScaleByDistanceProperty(expected);
            }

            AssertPropertyJson(LabelCesiumWriter.PixelOffsetScaleByDistancePropertyName, expected);
        }

        [Test]
        public void TestDistanceDisplayCondition()
        {
            var expected = new Bounds(10.0, 20.0);

            using (var packet = OpenPacket())
            using (var label = packet.OpenLabelProperty())
            using (var interval = label.OpenInterval())
            {
                interval.WriteDistanceDisplayConditionProperty(expected);
            }

            AssertPropertyJson(LabelCesiumWriter.DistanceDisplayConditionPropertyName, expected);
        }

        [Test]
        public void TestDisableDepthTestDistance()
        {
            const double expected = 1234.0;

            using (var packet = OpenPacket())
            using (var label = packet.OpenLabelProperty())
            using (var interval = label.OpenInterval())
            {
                interval.WriteDisableDepthTestDistanceProperty(expected);
            }

            AssertPropertyJson(LabelCesiumWriter.DisableDepthTestDistancePropertyName, expected);
        }

        [Test]
        public void TestScaleByDistance()
        {
            var expected = new NearFarScalar(100.5, 1.5, 200.5, 2.5);

            using (var packet = OpenPacket())
            using (var label = packet.OpenLabelProperty())
            using (var interval = label.OpenInterval())
            {
                interval.WriteScaleByDistanceProperty(expected);
            }

            AssertPropertyJson(LabelCesiumWriter.ScaleByDistancePropertyName, expected);
        }

        [Test]
        public void TestText()
        {
            const string expected = "asdf";

            using (var packet = OpenPacket())
            using (var label = packet.OpenLabelProperty())
            using (var interval = label.OpenInterval())
            {
                interval.WriteTextProperty(expected);
            }

            AssertPropertyJson(LabelCesiumWriter.TextPropertyName, expected);
        }

        [Test]
        public void TestTextInterval()
        {
            var startDate = new GregorianDate(2012, 4, 2, 12, 0, 0).ToJulianDate();
            var stopDate = new GregorianDate(2012, 4, 2, 12, 1, 0).ToJulianDate();

            var interval1Start = startDate;
            var interval1Stop = startDate.AddSeconds(1);
            var interval2Start = interval1Stop;
            var interval2Stop = startDate.AddSeconds(2);
            var interval3Start = interval2Stop;
            var interval3Stop = stopDate;

            const string interval1Value = "A";
            const string interval2Value = "B";
            const string interval3Value = "C";

            using (var packet = OpenPacket())
            using (var label = packet.OpenLabelProperty())
            using (var text = label.OpenTextProperty())
            using (var textIntervals = text.OpenMultipleIntervals())
            {
                using (var interval = textIntervals.OpenInterval(interval1Start, interval1Stop))
                {
                    interval.WriteString(interval1Value);
                }

                using (var interval = textIntervals.OpenInterval(interval2Start, interval2Stop))
                {
                    interval.WriteString(interval2Value);
                }

                using (var interval = textIntervals.OpenInterval(interval3Start, interval3Stop))
                {
                    interval.WriteString(interval3Value);
                }
            }

            AssertPropertyJson(LabelCesiumWriter.TextPropertyName, new List<Dictionary<string, object>>
            {
                new Dictionary<string, object>
                {
                    { "interval", CesiumFormattingHelper.ToIso8601Interval(interval1Start, interval1Stop, Iso8601Format.Compact) },
                    { StringCesiumWriter.StringPropertyName, interval1Value },
                },
                new Dictionary<string, object>
                {
                    { "interval", CesiumFormattingHelper.ToIso8601Interval(interval2Start, interval2Stop, Iso8601Format.Compact) },
                    { StringCesiumWriter.StringPropertyName, interval2Value },
                },
                new Dictionary<string, object>
                {
                    { "interval", CesiumFormattingHelper.ToIso8601Interval(interval3Start, interval3Stop, Iso8601Format.Compact) },
                    { StringCesiumWriter.StringPropertyName, interval3Value },
                },
            });
        }

        [Test]
        public void TestShowBackground()
        {
            const bool expected = true;

            using (var packet = OpenPacket())
            using (var label = packet.OpenLabelProperty())
            using (var interval = label.OpenInterval())
            {
                interval.WriteShowBackgroundProperty(expected);
            }

            AssertPropertyJson(LabelCesiumWriter.ShowBackgroundPropertyName, expected);
        }

        [Test]
        public void TestBackgroundColor()
        {
            var expected = Color.Red;

            using (var packet = OpenPacket())
            using (var label = packet.OpenLabelProperty())
            using (var interval = label.OpenInterval())
            {
                interval.WriteBackgroundColorProperty(expected);
            }

            AssertPropertyJson(LabelCesiumWriter.BackgroundColorPropertyName, expected);
        }

        [Test]
        public void TestBackgroundPadding()
        {
            var expected = new Rectangular(3, 4);

            using (var packet = OpenPacket())
            using (var label = packet.OpenLabelProperty())
            using (var interval = label.OpenInterval())
            {
                interval.WriteBackgroundPaddingProperty(expected);
            }

            AssertPropertyJson(LabelCesiumWriter.BackgroundPaddingPropertyName, expected);
        }

        [Test]
        public void TestTranslucencyByDistance()
        {
            var expected = new NearFarScalar(100.0, 1.0, 200.0, 2.0);

            using (var packet = OpenPacket())
            using (var label = packet.OpenLabelProperty())
            using (var interval = label.OpenInterval())
            {
                interval.WriteTranslucencyByDistanceProperty(expected);
            }

            AssertPropertyJson(LabelCesiumWriter.TranslucencyByDistancePropertyName, expected);
        }

        [Test]
        public void TestTranslucencyByDistanceSamples()
        {
            var epoch = new GregorianDate(2012, 4, 2, 12, 0, 0).ToJulianDate();

            using (var packet = OpenPacket())
            using (var label = packet.OpenLabelProperty())
            using (var interval = label.OpenInterval())
            using (var scaleByDistance = interval.OpenTranslucencyByDistanceProperty())
            {
                var dates = new List<JulianDate>();
                var values = new List<NearFarScalar>();

                dates.Add(epoch);
                values.Add(new NearFarScalar(100.0, 1.0, 200.0, 2.0));

                dates.Add(epoch.AddSeconds(60.0));
                values.Add(new NearFarScalar(200.0, 1.0, 300.0, 2.0));

                scaleByDistance.WriteNearFarScalar(dates, values);
            }

            Assert.AreEqual("{\"label\":{\"translucencyByDistance\":{\"epoch\":\"20120402T12Z\",\"nearFarScalar\":[0,100,1,200,2,60,200,1,300,2]}}}", StringWriter.ToString());
        }

        private void AssertPropertyJson(string propertyName, object value)
        {
            AssertExpectedJson(PacketCesiumWriter.LabelPropertyName, new Dictionary<string, object>
            {
                { propertyName, value },
            });
        }

        protected override CesiumPropertyWriter<LabelCesiumWriter> CreatePropertyWriter(string propertyName)
        {
            return new LabelCesiumWriter(propertyName);
        }
    }
}
